pkg/plugins/openai/cmd/main/main.go

```diff
@@ -0,0 +1,345 @@
+			tokenCount = -1
+		}
+
+		extendedAttrs := pb.CustomCostExtendedAttributes{
+			AccountId:    &billingEntry.OrganizationID,
+			SubAccountId: &billingEntry.ProjectID,
+		}
+		customCost := pb.CustomCost{
+			BilledCost:         float32(billingEntry.CostInMajor),
+			AccountName:        billingEntry.OrganizationName,
+			ChargeCategory:     "Usage",
+			Description:        fmt.Sprintf("OpenAI usage for model %s", billingEntry.Name),
+			ResourceName:       billingEntry.Name,
+			ResourceType:       "AI Model",
+			Id:                 uuid.New().String(),
+			ProviderId:         fmt.Sprintf("%s/%s/%s", billingEntry.OrganizationID, billingEntry.ProjectID, billingEntry.Name),
+			UsageQuantity:      float32(tokenCount),
+			UsageUnit:          "tokens - All snapshots, all projects",
+			ExtendedAttributes: &extendedAttrs,
+		}
+
+		customCosts = append(customCosts, &customCost)
+	}
+
+	return customCosts, nil
+}
+
+var snapshotRe = regexp.MustCompile(`-\d{4}-\d{2}-\d{2}|-`)
+
+func buildTokenMap(usage *openaiplugin.OpenAIUsage) map[string]int {
+	tokenMap := make(map[string]int)
+	if usage == nil {
+		return tokenMap
+	}
+	for _, usageData := range usage.Data {
+		key := snapshotRe.ReplaceAllString(usageData.SnapshotID, "")
+		key = strings.ToLower(key)
+		if _, ok := tokenMap[key]; !ok {
+			tokenMap[key] = 0
+		}
+
+		tokenMap[key] += (usageData.NGeneratedTokensTotal + usageData.NContextTokensTotal)
+	}
+	return tokenMap
+}
+
+func (d *OpenAICostSource) getOpenAIBilling(start time.Time, end time.Time) (*openaiplugin.OpenAIBilling, error) {
+	client := &http.Client{}
+	openAIBillingURL := fmt.Sprintf(openAIBillingURLFmt, start.Format(openAIAPIDateFormat), end.Format(openAIAPIDateFormat))
+	log.Debugf("fetching OpenAI billing data from %s", openAIBillingURL)
+	var errReq error
+	var resp *http.Response
+	for i := 0; i < 3; i++ {
+		err := d.rateLimiter.Wait(context.Background())
+		if err != nil {
+			log.Warnf("error waiting for rate limiter: %v", err)
+			return nil, fmt.Errorf("error waiting for rate limiter: %v", err)
+		}
+		var req *http.Request
+		req, errReq = http.NewRequest("GET", openAIBillingURL, nil)
+		if errReq != nil {
+			log.Warnf("error creating billing export request: %v", errReq)
+			log.Warnf("retrying request after 30s")
+			time.Sleep(30 * time.Second)
+			continue
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", d.config.APIKey))
+
+		resp, errReq = client.Do(req)
+		if errReq != nil {
+			log.Warnf("error doing billing export request: %v", errReq)
+			log.Warnf("retrying requestafter 30s")
+			time.Sleep(30 * time.Second)
+			continue
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			errReq = fmt.Errorf("received non-200 response for billing export request: %d", resp.StatusCode)
+			log.Warnf("got non-200 response for billing export request: %d", resp.StatusCode)
+			log.Warnf("retrying requestafter 30s")
```

Comment:
Would it be valuable to add the response's body to the error/logs?

Reply:
yup I like it